### DISCUSSION TITLE
Cursor mode

### DISCUSSION THREAD

**Discussion by Zaplae** [helix-editor/helix#10219] [Ideas]
# Cursor mode
I think the current multi-cursor system is not powerful and flexible enough.

<space + c> for `cursor mode`:

```
Cursor
n          Add <n> new cursor below. "<n>C" for short (<space> + c + <n> + n + Esc )
tab        Select <n> next real cursor
shiftab    Select <n> prev real cursor
Return     Make the next cursor become real
c          Change the kind of current cursor group. The following argument : `r` for `real`, `n` for `next`.
,          Remove all real cursor except the current selected real cursor
=          Remove next cursor and return to the position of the selected real cursor
```
Note: This mode should be continued and display this infobox until pressing `Esc` or `,`

There are two kinds of cursors, one is `next cursor` and the other is `real cursor`.

`next cursor` is single and only, always overlaps the selected `true cursor`. It represent the next real cursor.
`next cursor` didn't exist yet before we adjust its position and press `Return` to make it become `real cursor`.

A new  `next cursor` will be spawned when we press `Return` to make the last `next cursor` become real.(The new also overlaps)

If we want to move the selected real cursor, we could type `c` to change the current kind of cursor groups.
For example: We want to move a `real key` but not a `next key`,we should enter cursor mode and press `c` 
Default cursor group is `next cursor group`, which means we can only move `next cursor` by default
Default selected real cursor is that single real cursor

Note: 
You can restart by pressing `=`, 
You can move cursor just like normal mode.
- - -

## Example 1
If we want to select and edit `line 6` and `line 10`:
![image](https://user-images.githubusercontent.com/98511527/151400605-f79946f4-39e9-4e7c-b843-18a1408ca84f.png)
1. enter cursor mode
2. `:g 10` or `4j` to move the `next cursor` which overlapped the selected real cursor originally, then `Return` to ensure the `next cursor` which have moved to `line 10` will become real, and `Esc` return to normal mode
3. just `xc` to edit them

## Example 2
Select some words quickly and intuitively:
![image](https://user-images.githubusercontent.com/98511527/151402576-3e6498bf-7d14-41d0-b793-8afbfbbe7877.png)
1. enter cursor mode
2. `f b` + `Return`, `f c` + `Return` to make 2 `next cursors` become real , then `Esc` to return to normal mode
3. ``w` to select the whole word, and edit them

Why not provide some command like `swap` ? 
It is helpful because we can use it in normal mode: `:swap 1,3` to swap `aaaa` and `cccc`quickly

**Comment by NNBnh** [comment]
Hmm, interesting idea. But consider change the default binding for this 2 command:

> ```
> tab        Select <n> next real cursor
> shiftab    Select <n> prev real cursor
> ```

to something else more ergonomic.

**Comment by Zaplae** [comment]
@NNBnh Or `[`  `]`? I'm not sure......

**Comment by NNBnh** [comment]
If i understand correctly:

> ```
> n          Add <n> new cursor below. "<n>C" for short (<space> + c + <n> + n + Esc )
> ```

`n` here mean "insert a number" and not the <kbd>N</kbd> key right?

If so we could map it to <kbd>N</kbd> and <kbd>Shift</kbd> + <kbd>N</kbd> .

**Comment by EpocSquadron** [comment]
I suggested an alternative approach to this same idea in #943

**Comment by Zoybean** [comment]
I think I don't quite understand all of it (mostly the `n` and `c` commands), so I tried writing out my understanding of how it would work, and I think I'm wrong on a few points; but if I am, I think I want to propose what I wrote as an alternative design, mostly because I like the simplicity of always having a single "next cursor" that moves around and is used to create new "real cursors".
Here's my interpretation / design:
- When you enter cursor mode, all your current cursors become "real cursors" and a "next cursor" is placed under the primary "real cursor".
- In cursor mode, there is always one "next cursor".
- In cursor mode, motions only affect the one "next cursor" rather than all the "real cursors" (in normal or select mode it would affect all the "real cursors").
- Pressing `Return` creates a new "real cursor" at the current position of the "next cursor", and keeps the "next cursor" at that position.
- Pressing `=` moves "next cursor" back to the the current "real cursor".
- Pressing `Tab` or `Shift-Tab` cycles which "real cursor" is the current one, and moves the "next cursor" to the new current "real cursor", similar to changing the primary selection with `(` and `)`. (idea: Why not just use `(` and `)` for that? I don't think there's anything better for those keys to do in cursor mode)

If that's all correct, then I'm not sure I understand the `n` command or the `c` command. `n` seems like it creates multiple "next cursors", when there should only be 1. And I'm not sure what a "cursor group" would be for the `c` command.

Here's an example scenario just to check that I got the basics. If I had:
```
[s]ome words here
```
then did cursor mode + `w` + `Return`
```
[s]ome [w]ords here
```
then `w` + `Return` again
```
[s]ome [w]ords [h]ere
```
Is that correct?

**Comment by EpocSquadron** [comment]
@Zoybean I think what you wrote up might be closer to my alternative in #943. I'm a little foggy on this version from the original description though.
